Rnmobile/add/typography controls

enejb opened the pull request:
## Description
This Pr attempts to add the typographical controls needed to manipulate the font size and line height. 

## How has this been tested?
Prereqisits. 
1. WordPress 5.8. 
2. Gutenberg 11.1 
3. Theme that supports full site editing. (TT1 Blocks) for example.
4. Latest develop branch of iOS or Android. 

## Screenshots <!-- if applicable -->

https://user-images.githubusercontent.com/115071/126539233-bc06ff84-7f75-4fc8-8e5f-dd22595af0e1.mov


## Types of changes
New Feature for mobile app: Adds font size and line height controls blocks that have it.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

enejb committed:
Allow for a non units to show without the unit picker.

enejb committed:
Add Line Height Controls

enejb committed:
Make variable more clear

enejb committed:
Revert Typography web

enejb committed:
Add support for font sizes

enejb committed:
Add bottom Cell sheet subLabel

enejb committed:
Update the colour of the subheading

enejb committed:
Update colours again for subheading

enejb committed:
Make the Typography UI only available in dev build

enejb committed:
Do not show items that we can't parse as numbers

enejb committed:
Add comment

geriux commented on packages/components/src/index.native.js:
```diff
@@ -15,6 +15,7 @@ export { default as Dashicon } from './dashicon';
 export { default as Dropdown } from './dropdown';
 export { default as DropdownMenu } from './dropdown-menu';
 export { default as FocalPointPicker } from './focal-point-picker';
+export { default as FontSizePicker } from './font-size-picker';
```
On Android I'm getting this error:

```javascript
 ERROR  TypeError: Cannot read property 'createSlotFill' of undefined, js engine: hermes
```
I think it's because the `FontSizePicker` import is before:

```javascript
export {
	createSlotFill,
	Slot,
	Fill,
	Provider as SlotFillProvider,
} from './slot-fill';
```

geriux commented on packages/rich-text/src/component/index.native.js:
```diff
@@ -1070,24 +1079,48 @@ const withFormatTypes = ( WrappedComponent ) => ( props ) => {
+		const settings = getSettings();
+
+		const baseGlobalStyles = settings?.__experimentalGlobalStylesBaseStyles;
+		const fontSizes =
+			settings?.__experimentalFeatures?.typography?.fontSizes?.theme ??
+			settings.fontSizes ??
+			[];
+
+		const currentBlockAttributes = getBlockAttributes( clientId );
+		const blockFontSize = currentBlockAttributes?.fontSize;
+		const blockStyle = currentBlockAttributes?.style;
+
+		const fontSizeObject = getBlockFontSize(
+			fontSizes,
+			blockFontSize,
+			blockStyle?.typography?.fontSize
+		);
+		const blockFontSizeValue =
+			fontSizeObject?.size ||
+			blockStyle?.typography?.fontSize ||
+			blockFontSize;
```
I wonder if we should add this in a separate PR, this way this one would be mainly the UI what do you think?

enejb replied:
Yes, I will remove this part since it doesn't quite work as expected.

geriux commented on packages/block-editor/src/hooks/typography.native.js:
```diff
@@ -0,0 +1,64 @@
+
+	// only enable TypographyPanel for development
+	// eslint-disable-next-line no-undef
+	if ( isDisabled || ! isSupported || ! __DEV__ ) return null;
```
Nice! 👏

geriux commented:
Can we also include in this PR the removal of this [line](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/hooks/typography.js#L87
)?

Since we are now adding a `native` variant, that check won't be needed.

geriux commented on packages/components/src/font-size-picker/index.native.js:
```diff
@@ -0,0 +1,173 @@
+					</BottomSheet.Cell>
+					{ fontSizes.map( ( item, index ) => {
+						// Only display a choice that we can currenly select.
+						if ( ! parseFloat( item.size ) ) {
+							return null;
+						}
```
Nice check!

geriux commented on packages/components/src/font-size-picker/index.native.js:
```diff
@@ -0,0 +1,173 @@
+								accessibilityLabel={
+									item.size === selectedValue
+										? sprintf(
+												// translators: %s: Select control option value e.g: "Auto, 25%".
```
Should we update this? So it has more context with the font size picker.

enejb replied:
Good idea!

geriux commented:
Hey @enejb can you please update this PR with the latest `trunk`? I couldn't test on Android even after handling the [import issue](https://github.com/WordPress/gutenberg/pull/33605#discussion_r690978899). I think some code changes a missing like [this one](https://github.com/WordPress/gutenberg/pull/33707).

enejb committed:
Move the FontSizePicker export

enejb committed:
Remove the not needed platform check

enejb committed:
Update comment to be more font size specific.

enejb commented:
@geriux I removed the commits that were related to the font size changing and rebased the PR. 
Now that the code is behind the dev flag I think we can merge it continue working on the other parts. 
Such as font resizing and line height resizing.

enejb committed:
Fix unit test for customLineHeight

geriux commented on packages/components/src/unit-control/index.native.js:
```diff
@@ -115,19 +116,20 @@ function UnitControl( {
 		return (
 			<View style={ styles.unitMenu } ref={ anchorNodeRef }>
 				{ renderUnitButton }
-				{ hasUnits( units ) && units?.length > 1 ? (
```
I think we need to keep the `units?.length > 1`

Introduced from this [PR](https://github.com/WordPress/gutenberg/pull/33641).

enejb replied:
for sure make sense to me!

geriux commented:
Thanks for the changes @enejb! All is working correctly now on both iOS and Android 🎉 

I only have one [question](https://github.com/WordPress/gutenberg/pull/33605#discussion_r692801615) and then we can merge this 🚀

enejb committed:
Revert the units logic to hide the Unit picker

enejb committed:
Remove the added borderRightWidth for picker that doesn't have childern

enejb commented:
I addressed the issue and make it look a little nicer on iOS (removed the extra right border that I found). 
Can you take another look?

geriux reviewed (approved):
LGTM! Great work @enejb 👏 I tested it on both iOS and Android.
